File: 1963-minimum-number-of-swaps-to-make-the-string-balanced/1963-minimum-number-of-swaps-to-make-the-string-balanced.cpp
class Solution {
public:
    int minSwaps(string s) {
        int start = 0, end = s.size() - 1;
        int cnt = 0;

        stack<int> st;
        while (start <= end) {
            if (s[start] == ']') {
                if (st.empty()) {
                    while (s[end] == ']') end--;
                    s[start] = '[';
                    s[end] = ']';
                    cnt++;
                    st.push('[');
                }
                else st.pop();
            }
            else
                st.push('[');
            start++;
        }
        return cnt;
    }
};

File: 1963-minimum-number-of-swaps-to-make-the-string-balanced/1963-minimum-number-of-swaps-to-make-the-string-balanced.java
class Solution {
    public int minSwaps(String s) {
        int openCnt = 0;

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == ']') {
                if (openCnt == 0)
                    openCnt++;
                else
                    openCnt--;
            }
            else openCnt++;
        }
        return (openCnt + 1) / 2;
    }
}